### PR TITLE
RFC for TensorflowData -- Java API

### PR DESCRIPTION
**Pull request by dhruvrajan**

### CODE REVIEW

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+[tf.data](https://www.tensorflow.org/api_docs/python/tf/data).
+
+## Motivation
+
+A large part of the standard deep-learning workflow is managing datasets:
+configuring iteration and batching over a sequence of elements,
+and performing transformations on individual or grouped dataset elements.
+
+The goal of this project is to provide a high-level abstraction which simplifies
+this workfow, based on TensorFlow data ops, with syntax closely mirorring
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: typo in "workflow" and in "mirroring"

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    {7, 8, 9},
+    {10, 11, 12}
+}
+
+float[] labels = new float[] {
+    0,
+    1,
+    1,
+    0
+}
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: Would be interesting to show the examples using TF Tools, I can help you out with this

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+);
+```
+
+
+Other data sources are also possible, using `tf.data` ops. Datasets can also be constructed from
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: truncated sentence?

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+dataset = dataset.batch(2)
+```
+
+Dataset transformations alter both the values and shapes of the original elements, and
+return a *new* `Dataset` object.
+
+In this case, the original dataset had 4 elements of shape `[features: (3,) labels: (1,)]`.
+Once the `.batch` transformation is applied, the new dataset has 2 elements (batches) of shape `[features: (2, 3), labels: (2, 1)]`.
+
+Similar transformations include `.skip`, `.take`, `.map`, `.filter`, etc.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md:109]: cool stuff!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+
+### Iterating over Dataset Elements
+
+The primary use of a dataset is for iteration over its elements.
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this elements can be accessed using `List.get(int index)`.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md:116]: `List.get(int)` will return you a tensor, not an output, no? 

Maybe it would be interesting to leave the batch items as `Operand` or `Constant` objects? `Output` became in time kind of a more internal data class.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+A dataset can be constructed from a list of the constant `Operand`s generated
+from this dataset, and a list of `DataType` objects corresponding
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md:86]: or could we do something like:
```java
Dataset.fromTensorSlices(tf, Arrays.asList(features, labels), ...)
```
and let the `Dataset` create the constant? Also, the datatypes might became optional since often then can be inferred from the type of the constant.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this elements can be accessed using `List.get(int index)`.
+
+It is recommended to use `Tensor.expect(DataType<?> dtype)` to restore types
+to the retrieved tensors.
+
+
+#### Eager Mode: Iterable
+The `Dataset` class implements the `Iterable` interface, so in
+eager mode, iteration over dataset elements is possible using a standard for-each loop.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: that could still be useful in graph mode, the only difference is that you won't be able to access the tensor or operate directly on the outputs in your iteration. But you could do other stuff with the operands.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+// Dataset of (features, labels) from above
+Dataset dataset = Dataset.fromTensorSlices(tf, ... );
+
+// batch dataset elements into batches of size 2
+dataset = dataset.batch(2); 
+
+for (List<Output<?>> batch : dataset) {
+    Tensor<TInt32> featureBatch = element.get(0).tensor().expect(TInt32.DTYPE);
+    Tensor<TInt32> labelBatch = element.get(1).tensor().expect(TInt32.DTYPE);
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: Actually, if elements have already been transformed into a constant, like you were suggesting, then it is better to keep them as constant for operating them in eager mode as well, so the example should illustrate this.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+    // Perform batch-wise computations on featreBatch and labelBatch
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: typo in "featureBatch"

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+#### Graph Mode: OneShotIterator
+
+The above code will not work in graph mode, which requires the use of `Session`s
+to run the computations. In graph mode, datasets can be iterated over using the `OneShotIterator` abstraction, and a while loop.
+
+`OneShotIterator` allows iteration over a dataset's elements "one time", using
+repeated calls to `Session.run`. 
+
+The `OneShotIterator` object has two fields:
+- `OneShotIterator.getMakeIteratorOp()` returns a "make-iterator" `Operation` which is run to initialize this iterator
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: might be interesting to see where that will fit with changes from Adam, he add something related to graph initialization run.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+            try {
+                List<Tensor<?>> outputs = session.runner()
+                    .fetch(batch.get(0))
+                    .fetch(batch.get(1))
+                    .run();
+
+                Tensor<TInt32> featureBatch = outputs.get(0).expect(TInt32.DTYPE);
+                Tensor<TInt32> labelBatch = outputs.get(1).expect(TInt32.DTYPE);
+
+                // Perform batch-wise computations on featreBatch and labelBatch
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: typo in "featureBatch"

File: sigs/jvm/rfcs/20200126-tensorflow-data.md [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+            }
+        }
+    }
+}
+
+```
+
+## Questions and Discussion Topics
+
+Se [MNISTBasicEagerClassifier.java](https://github.com/dhruvrajan/tensorflow-java/blob/tensorflow-keras-dev/tensorflow-frameworks/tensorflow-keras/src/main/java/org/tensorflow/keras/examples/mnist/MNISTBasicEagerClassifier.java) and [MNISTBasicGraphClassifier.java](https://github.com/dhruvrajan/tensorflow-java/blob/tensorflow-keras-dev/tensorflow-frameworks/tensorflow-keras/src/main/java/org/tensorflow/keras/examples/mnist/MNISTBasicGraphClassifier.java) as examples using the described API
```
Comment (karllessard) [inline sigs/jvm/rfcs/20200126-tensorflow-data.md]: typo in "See"

**Review by karllessard — COMMENTED** [review]
I think it is an interesting proposal! I did something more basic to write a MNIST example I would like to publish soon too: 
https://github.com/karllessard/tensorflow-java-models/tree/master/tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/data

I would gladly try to replace this with your datasets when they'll be ready

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 1 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: thanks!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 9 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: thanks!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 12 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: thanks!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 11 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: thanks!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 2 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: Ah, absolutely!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 3 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: updated this!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 5 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md:116]: This method could return `List<Operand<?>>` (can't return a `Constant`, I think, since these are outputs of tf.data ops).

I think users won't often have to work with `Tensor` very often here, since these data values only need to be passed to other operations, in most workflows.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 6 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md:86]: Inferring the type automatically here would be cool!

Keeping this interface  (as well) might be useful though, in case users have tensorflow operations they want to run before constructing the dataset? That'd be a bit strange, but possible, I suppose...

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 7 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: Hmm; wondering what the semantics of this would be. In graph mode, you don't end up getting a new `List<Output<?>>` for each "batch"; you use one `List<Output<?>>`, which is automatically updated from the original `Constant` tensors on successive `Session.run` calls.

so allowing:

```
for (List<Output<?>> : graphDataset);
```
would be a bit strange?

Would this be more like a `map` that operates over individual elements? We can allow

```
graphDataset.map(Function<List<Output<?>>, List<Output<?>>>);
```

which would probably accomplish the same thing?

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 8 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: I agree, I just updated the examples here!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 10 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: Yup, I added a `DatasetIterator` to replace the `OneShotIterator` class (more closely mimics the `tf.data.Iterator` from Python too). It uses the new `tf.initAdd` / `tf.init`!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 7 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: I think it's best to keep the `while` loop, as it's consistent with TF Python's API, and other semantics start to get really strange, since it would hide a `session.run` call.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 5 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md:116]: We could change this to be a `List<Operand<?>>`, if we modify the core `tf.data.iteratorGetNextOp`.

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 2 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md]: I updated these examples to use a FloatNdArray; the array creation is about the only place `NdArray` functionality is used.

**Comment by dhruvrajan** [comment]
I've updated this RFC with changes that make the iteration logic a lot simpler!

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 6 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md:86]: Also, I think there aren't `tf.constant` signatures which allow input of `NdArray<?>`,  so it might be tough to preserve the ability to input tensors with different dtypes.

We could add multiple typed factories here, though, for `IntNdArray`, `FloatNdArray`, etc...

**Comment by karllessard** [comment]
I just saw that you added some comments in this RFC as well as in the PR. If your PR is closed to be ready for merge, I suggest that we skip the RFC and continue the discussion exclusively on the PR, wdyt?

File: sigs/jvm/rfcs/20200126-tensorflow-data.md (hunk 6 above)
Comment (karllessard) [reply sigs/jvm/rfcs/20200126-tensorflow-data.md:86]: mmh, there are... for example, [this one](https://github.com/tensorflow/java/blob/9fe9d87a454308e020b57839abf2e061f04c5a1e/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L1363). That is what you were looking for, right?

Oh you mean for the untyped `NdArray<?>` explicitly? Ok, yeah I skipped that one, which would conflict anyway with `NdArray<String>`. I thought that while tensor factories should allow the creation of tensor of any type, the "shortcut" constant allocation are more restrained to a finite list of popular datatypes. If your datatype is not supported, then you end up using [`constant(Tensor<T>)`](https://github.com/tensorflow/java/blob/9fe9d87a454308e020b57839abf2e061f04c5a1e/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L1691) directly (but that forces you to close yourself the tensor after the constant is created).

I'm not sure if that is the right way of doing it neither, I'm open to suggestions!

**Comment by dhruvrajan** [comment]
@karllessard I think that makes sense! I have added the information in this RFC to the README of the Data PR.

**Comment by dhruvrajan** [comment]
closing in favor of the PR.
